lib/collection/src/shards/transfer/shard_transfer.rs

```diff
@@ -773,67 +760,72 @@ where
     F: Future<Output = ()> + Send + 'static,
 {
     spawn_async_cancellable(move |cancel| async move {
-        let mut tries = MAX_RETRY_COUNT;
-        let mut finished = false;
-        while !finished && tries > 0 {
-            let transfer_result = transfer_shard(
-                transfer.clone(),
-                shards_holder.clone(),
-                consensus.as_ref(),
-                collection_id.clone(),
-                &collection_name,
-                channel_service.clone(),
-                &snapshots_path,
-                &temp_dir,
-                cancel.clone(),
-            )
-            .await;
-            finished = match transfer_result {
-                Ok(()) => true,
-                Err(error) => {
-                    // Revert queue proxy if we still have any to clean up or prepare for the next attempt
-                    if let Some(replica_set) =
-                        shards_holder.read().await.get_shard(&transfer.shard_id)
-                    {
-                        replica_set.revert_queue_proxy_local().await;
-                    }
+        let mut result = Err(CollectionError::Cancelled {
+            description:
+                "shard transfer task exit without attempting any work, this is a programming error"
+                    .into(),
+        });
 
-                    if matches!(error, CollectionError::Cancelled { .. }) {
-                        return false;
-                    }
+        for attempt in 0..MAX_RETRY_COUNT {
+            let future = async {
+                if attempt > 0 {
+                    sleep(RETRY_DELAY * attempt as u32).await;
+
+                    log::warn!(
+                        "Retrying shard transfer {} -> {} (retry {attempt})",
+                        transfer.shard_id,
+                        transfer.to,
+                    );
+                }
+
+                transfer_shard(
+                    transfer.clone(),
+                    shards_holder.clone(),
+                    consensus.as_ref(),
+                    collection_id.clone(),
+                    &collection_name,
+                    channel_service.clone(),
+                    &snapshots_path,
+                    &temp_dir,
+                )
+                .await
+            };
+
+            result = cancel::future::cancel_on_token(cancel.clone(), future)
+                .await
+                .map_err(Into::into)
+                .and_then(|res| res);
+
+            match &result {
+                Ok(()) => break,
+
+                // TODO: Should we break early if shard transfer is cancelled, or should we handle it as regular error?
```

Comment:
We want to handle it as regular error. More specifically, we want to call `revert_queue_proxy_local()`. We probably call this later as well as part of aborting the snapshot, but doing it here as well is a good thing to be sure.

Reply:
What about changing this `match` to:

```rust
match &result {
    Ok(()) => break,

    Err(err) => {
        // Revert queue proxy if we still have any to clean up
        if let Some(shard) = shards_holder.read().await.get_shard(&transfer.shard_id) {
            shard.revert_queue_proxy_local().await;
        }

        if let CollectionError::Cancelled { .. } = err {
            break;
        }

        log::error!(
            "Failed to transfer shard {} -> {}: {err}",
            transfer.shard_id,
            transfer.to,
        );
    },
}
```

Reply:
Ooooh, that's a nice suggestion, I like it. 👌

Reply:
It does reorder logging after reverting the queue proxy, but I guess it's fine.

Reply:
Now that I think of it, we can probably remove the whole match and do an 'early return' if `Ok(())`` to clean this up even further.

Reply:
I've restructured `spawn_transfer_task`, what do you think? 🤔